File: src/main/java/com/github/bjoernpetersen/jmusicbot/provider/Provider.java
package com.github.bjoernpetersen.jmusicbot.provider;

import com.github.bjoernpetersen.jmusicbot.IdPlugin;
import com.github.bjoernpetersen.jmusicbot.InitStateWriter;
import com.github.bjoernpetersen.jmusicbot.InitializationException;
import com.github.bjoernpetersen.jmusicbot.PlaybackFactoryManager;
import com.github.bjoernpetersen.jmusicbot.Song;
import com.github.bjoernpetersen.jmusicbot.playback.PlaybackFactory;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import javax.annotation.Nonnull;

public interface Provider extends IdPlugin {

  Set<Class<? extends PlaybackFactory>> getPlaybackDependencies();

  /**
   * Gets the base class for this provider. The base class should define relevant methods for {@link
   * Suggester} implementations depending on this provider.
   *
   * This provider must implement the returned base class.
   *
   * @return a class
   */
  @Nonnull
  Class<? extends Provider> getBaseClass();

  /**
   * Signals to the provider that it can start doing work.
   *
   * @param manager a PlaybackFactoryManager to obtain PlaybackFactory instances from
   * @param initStateWriter a writer for initialization state messages
   * @throws InitializationException If any errors occur (for example due to invalid log-in
   * credentials)
   * @throws InterruptedException if the thread is interrupted while initializing
   */
  void initialize(@Nonnull InitStateWriter initStateWriter, @Nonnull PlaybackFactoryManager manager)
      throws InitializationException, InterruptedException;

  /**
   * Searches for songs based on the given search query.
   *
   * It is recommended only to return about 30 songs.
   *
   * @param query a search query, trimmed and not empty
   * @return a list of songs
   */
  @Nonnull
  List<Song> search(@Nonnull String query);

  /**
   * Looks up a song by its ID.
   *
   * @param id the song ID
   * @return the song with the specified ID
   * @throws NoSuchSongException if the ID is invalid
   */
  @Nonnull
  Song lookup(@Nonnull String id) throws NoSuchSongException;

  /**
   * Looks up a batch of song IDs. If any can't be looked up, they will be dropped.
   *
   * @param ids a list of song IDs
   * @return a list of songs
   */
  @Nonnull
  default List<Song> lookupBatch(@Nonnull List<String> ids) {
    List<Song> result = new ArrayList<>(ids.size());
    for (String id : ids) {
      try {
        result.add(lookup(id));
      } catch (NoSuchSongException e) {
        // TODO log or something
      }
    }
    return Collections.unmodifiableList(result);
  }
}

File: src/main/kotlin/com/github/bjoernpetersen/musicbot/spi/plugin/Provider.kt
package com.github.bjoernpetersen.musicbot.spi.plugin

import com.github.bjoernpetersen.musicbot.api.Song
import java.io.IOException
import kotlin.reflect.KClass

interface Provider : Plugin, UserFacing {

    /**
     * Searches for songs based on the given search query.
     *
     * It is recommended only to return about 30 songs at max.
     *
     * @param query a search query, trimmed and not empty
     * @return a list of songs
     */
    fun search(query: String): List<Song>

    /**
     * Looks up a song by its ID.
     *
     * @param id the song ID
     * @return the song with the specified ID
     * @throws NoSuchSongException if the ID is invalid
     */
    @Throws(NoSuchSongException::class)
    fun lookup(id: String): Song

    fun getPlaybackSupplier(song: Song): PlaybackSupplier

    /**
     * Loads a song, i.e. prepares it before it can be played.
     *
     * This method can potentially download a song before it is being played. The song might never be
     * actually played, so this should not prepare a {@link Playback} object in any way.
     *
     * This method will not be called on an UI Thread, so it may block.
     *
     * @param song the song to load
     * @return whether the song has been loaded successfully
     */
    fun loadSong(song: Song): Boolean

    /**
     * Looks up a batch of song IDs. If any can't be looked up, they will be dropped.
     *
     * @param ids a list of song IDs
     * @return a list of songs
     */
    fun lookupBatch(ids: List<String>): List<Song> {
        val result = ArrayList<Song>(ids.size)
        for (id in ids) {
            try {
                result.add(lookup(id))
            } catch (e: NoSuchSongException) {
                // TODO log or something
            }
        }
        return result.toList()
    }
}

interface PlaybackSupplier {
    /**
     * Supplies the playback object for the specified song.
     *
     * @param song a song
     * @return a Playback object
     * @throws IOException if the playback could not be created
     */
    @Throws(IOException::class)
    fun supply(song: Song): Playback
}

fun playbackSupplier(supplier: (Song) -> Playback) = object : PlaybackSupplier {
    override fun supply(song: Song): Playback = supplier(song)
}

class NoSuchSongException : Exception {
    /**
     * Specify the ID of the missing song.
     * @param id a song ID
     */
    constructor(id: String) : super("Could not find song with ID $id")

    /**
     * Specify the ID of the missing song and the most specific base of the failed provider.
     * @param id a song ID
     * @param providerBase the most specific base of the throwing provider
     */
    constructor(id: String, providerBase: KClass<out Provider>) :
        super("Provider ${providerBase.qualifiedName} could not find song with ID $id")
}
